### PR TITLE
Enforce HTTP request Content-Length correctness

### LINKED ISSUE #62258: SocketsHttpHandler does not enforce request Content-Length correctness
The Content-Length header we send may not match the number of bytes actually sent.
- If the custom content is smaller than Content-Length, this will result in the request waiting on either the client or server to time out. The cause of such a hang can be non-obvious (https://github.com/microsoft/reverse-proxy/issues/1412).
- If the content is larger, it may corrupt the connection for future requests.

`SocketsHttpHandler` does not currently enforce that the number of bytes sent matches the header. The only exception to that is that [HTTP/3 will not allow you to send too much content](https://github.com/dotnet/runtime/blob/7b0a6b50cbedb30bd243a836e2b223806f4fc072/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L415-L419).

Since this is a protocol violation and non-obvious to diagnose, I suggest that we always throw in case the Content-Length does not match the number of bytes sent. We already track the number of bytes sent (for telemetry purposes), it's just a matter of reacting to a mismatch.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #62258

Like is already the case for HTTP/3, throw on an attempt to write too much request content. This means the server will never receive more than `Content-Length` bytes.

Also throw if we end up sending too few bytes.

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs [hunk 1]
```diff
@@ -3365,4 +3365,57 @@ public sealed class SocketsHttpHandler_RequestValidationTest_Sync : SocketsHttpH
...
+                }
+                catch { }
+            });
+        }
+    }
+
+    public sealed class SocketsHttpHandler_RequestContentLengthMismatchTest_Http11 : SocketsHttpHandler_RequestContentLengthMismatchTest
+    {
+        public SocketsHttpHandler_RequestContentLengthMismatchTest_Http11(ITestOutputHelper output) : base(output) { }
+        protected override Version UseVersion => HttpVersion.Version11;
```
Comment (halter73) [inline src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:3404]: Does throwing from the `ContentLengthWriteStream` close the HTTP/1.1 connection? If so, can this be verified in the test?

Comment (MihaZupan) [reply src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:3404]: Yes, any exception thrown during the request content transfer will result in the connection being torn down.
I suppose we could explicitly test that as well

**Review by ManickaP — APPROVED** [review]
LGTM, thanks.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 2]
```diff
@@ -414,8 +422,7 @@ private async ValueTask WriteRequestContentAsync(ReadOnlyMemory<byte> buffer, Ca
 
                 if (buffer.Length > _requestContentLengthRemaining)
                 {
-                    string msg = SR.net_http_content_write_larger_than_content_length;
-                    throw new IOException(msg, new HttpRequestException(msg));
+                    throw new HttpRequestException(SR.net_http_content_write_larger_than_content_length);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:425]: Do we care about this being an `IOException`?
It's technically a combination of IO + wrong headers.

Currently you would get
```c#
new HttpRequestException(
    net_http_content_stream_copy_error,
    new IOException(
        net_http_content_write_larger_than_content_length,
        new HttpRequestException(
            net_http_content_write_larger_than_content_length)))
```
after this PR
```c#
new HttpRequestException(
    net_http_content_write_larger_than_content_length)
```

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:425]: I assume the change here is to provide the same exception chain as with H/2 and HTTP/1.1? If so, I don't see a reason against.

Comment (halter73) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:425]: I understand not wanting to change the outer exception type at this point, but why wouldn't this cause an `InvalidOperationException`? That's what gets thrown by Kestrel if you try to write too much to the response stream after setting a `Content-Length` header. To me, it indicates a bug in the program.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:425]: > why not change the outer exception type? I hope nothing is relying on HttpClient throwing an HttpRequestException

`HttpClient` will always try to throw either a `TaskCanceledException` or an `HttpRequestException`.
Aside from initial argument validation, If something else gets thrown somewhere, we'll wrap it in `HttpRequestException`.

Sending too much or too little is the same kind of issue - the header you gave us doesn't match the content. Ideally, the user should see a similar kind of error.

> why wouldn't this cause an InvalidOperationException? That's what gets thrown by Kestrel if you try to write too much to the response stream

Most `HttpClient` users won't actually do the `WriteAsync` that goes over the limit themselves. They will use the built-in `HttpContent` types.
```c#
var content = new StreamContent(myStream);
content.Headers.ContentLength = someContentLength;

await client.PostAsync("foo", content);
```

I wouldn't expect this to throw `InvalidOperationException` as a user. It's not something I would consider an invalid operation, but more an argument/input exception.
Similarly, I wouldn't expect this to throw something different if my content length calculation was off in the other direction.

**Pull request merged**